src/libraries/Common/tests/System/IO/ConnectedStreams.cs

```diff
@@ -0,0 +1,163 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+using System.IO;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO
+{
+    /// <summary>Provides support for in-memory producer/consumer streams.</summary>
+    internal sealed class ConnectedStreams
+    {
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional() =>
+            CreateBidirectional(StreamBuffer.DefaultInitialBufferSize);
+
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <param name="initialBufferSize">The initial buffer size to use when storing data in the connection.</param>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional(int initialBufferSize) =>
+            CreateBidirectional(initialBufferSize, int.MaxValue);
+
+        /// <summary>Creates a pair of streams that are connected for bidirectional communication.</summary>
+        /// <param name="initialBufferSize">The initial buffer size to use when storing data in the connection.</param>
+        /// <param name="maxBufferSize">
+        /// The maximum buffer size to use when storing data in the connection.  When this limit is reached,
+        /// writes will block until additional space becomes available.
+        /// </param>
+        /// <remarks>Writing to one stream produces data readable by the either, and vice versa.</remarks>
+        public static (Stream Stream1, Stream Stream2) CreateBidirectional(int initialBufferSize, int maxBufferSize)
+        {
+            var b1 = new StreamBuffer(initialBufferSize, maxBufferSize);
+            var b2 = new StreamBuffer(initialBufferSize, maxBufferSize);
+            StrongBox<int> refCount = new StrongBox<int>(2);
+
+            return (new StreamBufferStream(b1, b2, refCount), new StreamBufferStream(b2, b1, refCount));
+        }
+
+        private sealed class StreamBufferStream : Stream
+        {
+            private readonly StreamBuffer _readBuffer;
+            private readonly StreamBuffer _writeBuffer;
+            private readonly StrongBox<int> _refCount;
+            private bool _disposed;
+
+            internal StreamBufferStream(StreamBuffer readBuffer, StreamBuffer writeBuffer, StrongBox<int> refCount)
```

Comment:
It would be good to have a comment about the need for StrongBox here, since this isn't obvious.

Reply:
Ok.